### ISSUE TITLE
[New Resource] AdapterName

### ISSUE THREAD

**Issue by PlagueHO** [PowerShell/xNetworking#115]
Implement a resource that is able to rename a network adapter based on specific query criteria. This resource should make use of the https://github.com/PowerShell/xNetworking/blob/dev/xNetworkAdapter.psm1 module to perform the Network Adapter query.

If additional adapter query parameters are required by this then the xNetworkAdapter.psm1 module should be extended to support the requirements.

See this issue for more information.
https://github.com/PowerShell/xNetworking/issues/94

**Comment by gbrayut** [comment]
Anyone working on this right now? I'm trying to migrate from our [StackExchange_NetworkAdapter](https://github.com/PowerShellOrg/StackExchangeResources/tree/master/DSCResources/StackExchange_NetworkAdapter) resource over to the xNetworking resource, but need to be able to update adapter names based on their description (example: `Intel(R) Gigabit 4P I350-t rNDC #2`).

Looks like the xNetworkAdapter.psm1 doesn't support using the description as a criteria yet, but I'm happy to add it. Gonna start working on a branch tomorrow.

**Comment by PlagueHO** [comment]
Awesome! I'm not working on this. That's good with me - @TravisEz13, @tysonjhayes ? Any help is most welcome :+1:

**Comment by TravisEz13** [comment]
Can you propose a schema?

**Comment by gbrayut** [comment]
[StackExchange_NetworkAdapter.schema.mof](https://github.com/PowerShellOrg/StackExchangeResources/blob/master/DSCResources/StackExchange_NetworkAdapter/StackExchange_NetworkAdapter.schema.mof) is the schema for our current DSC resource, where Name is the the desired name and Description is the unique Device Name of the network adapter (same as the Name property of Win32_NetworkAdapter WMI class). It also has TeamName and Ensure for setting up Team Nics, but I see that is handled by other resources in xNetworking. Looks like other xNetworking resources use State to toggle items, so I could see using that to enable/disable a network adapter. So I guess a proposed schema would be:

```

[ClassVersion("1.0.0"), FriendlyName("xNetAdapterName")]
class MSFT_xNetAdapterName : OMI_BaseResource
{
	[Key, Description("Desired name for the network adapter.")] string InterfaceAlias;
	[Key, Description("Device name of the network adapter.")] string DeviceName;
  	[write,Description("Specifies if the network adapter should be Enabled or Disabled."),ValueMap{"Enabled", "Disabled"},Values{"Enabled", "Disabled"}] string State;
};
```

We include the network information in our node data (See below) so ideally something like the above resource could be used to generate a name like "CO-AVSCAN01-PRI-VMNIC" using string concatenation and the xNetAdapterName resource.

```
@{
    Name = 'CO-AVSCAN01'
    Location = 'CO'
    ServerType = 'VM'
    NodeName = '3250cea4-7144-4dae-a5e6-1642bd2fa434'
    Network = @(
        @{
            Name = 'VMNIC'
            Zone = 'PRI'
            Description = 'vmxnet3 Ethernet Adapter'
        }
    )
}
```

**Comment by TravisEz13** [comment]
I don't see how this schema you can reliably rename a Network Adapter.  How do you identify the adapter to rename?

**Comment by gbrayut** [comment]
Device Name is a unique identifier for the network card. Duplicate devices get a "#2" or "#3" appended to the end of the device name.

**Comment by PlagueHO** [comment]
@gbrayut - so if the name param was being used to identify the adapter to rename,  what would we use to specify the new name of the adapter?

It would be possible to add a newname parameter to the resource. But that would expose another problem: once renamed the adapter would no longer be found.

What might work though would be to extend the xnetworkadapter to expose additional adapter properties that would be used as the resource key.  For example adapter driver,  mac address, description. You could then expose those parameters in the resource and use them as a compound key.

What I'm saying is that I think we'll run into problems in any resource if it renames a key property in the resource.

**Comment by gbrayut** [comment]
So I think there may be some confusion here. Just to make sure we are all on the same page let me clarify. I also attached a screenshot from a system with multiple NICs from the same vendor.

![image](https://cloud.githubusercontent.com/assets/304401/17465110/d1993e7a-5cab-11e6-9ec9-5f950e3de4ea.png)

DeviceName as used in the above schema refers to the unique system assigned description of the NIC driver, including the system assigned incremental digit at the end. This cannot be changed and is a good way to uniquely identify a network card. I highlighted this in red above, and it is also used as the Description field in `ipconfig /all`, the Name field in `gwmi Win32_NetworkAdapter | fl Name,AdapterType,DeviceID`, or the InterfaceDescription in ` gwmi MSFT_NetAdapter -Namespace ROOT/StandardCimv2 | fl Name,DriverDescription,InterfaceDescription`

InterfaceAlias as used in the above schema refers to the desired alias you want to assign to the network device. This is highlighted in red above, and is also used as the Name field in the two WMI queries and the section header in ipconfig (ex: "Ethernet adapter NY-RDP01-PRI-TEAM - VLAN 110"). Most of the other resources in this module already use this alias when selecting which device you want to assign an IP address, create a static route, or create a network team, as it is a cleaner and simpler way of selecting a device. In most systems these aliases default to "Ethernet", or "Ethernet 2", but the goal  of this new xNetAdapterName resource would be that you could assign the name as part of DSC (something we currently do using the StackExchange_NetworkAdapter resource). To prevent issues with the rename, I assume you would need to assign the xNetAdapterName instance as as dependency for any of the other networking DSC resources, or just know that it would generate an error until the new name was assigned.

It seems like InterfaceAlias is already used in these modules to refer to the user assigned name, but if the DeviceName, parameter is too confusing we could use InterfaceDescription instead. Also the above schema only supports setting the name for a given InterfaceDescription (works well in many cases since it should be known when the server is created), but I'd be happy to change that to an optional argument and allow selection via MAC address (note these also are technically user assignable and could change) or more generic criteria that is already accepted by the xNetworkAdapter.psm1 module.

**Comment by gbrayut** [comment]
So for example the description of the new DSC resource would be something like:

**xNetAdapterName**

- **InterfaceAlias**: Mandatory. Specifies the new alias name to assign to the selected network interface.
- **InterfaceDescription**: Optional, with no default. Specifies which network device should be renamed. Example "Intel(R) Gigabit 4P X540/I350 rNDC #2" or "vmxnet3 Ethernet Adapter".
- **MacAddress**: Optional, with no default. The MAC address of the network adapter you want to rename.

I am not sure what other criteria would make sense to select on. I guess a CurrentInterfaceAlias parameter might work if you know the current alias name is Ethernet or whatever , but it is a bit sloppy since after the rename you can't tell if the system is in the desired state (no way to know which interface the user was trying to select originally, you can only verify that the old name does not exist and the new name does). Adding PhysicalMediaType or Status might work, but since those are not guaranteed to be unique it may need another parameter like IgnoreMultipleMatchingAdapters, but that is also pretty sloppy. 

Seems like InterfaceDescription and MacAddress would be the best to start with since they are unique and can be verified easily.

**Comment by PlagueHO** [comment]
@gbrayut - doh! I apologize,  I misread your original MOF proposal. I put the blame on it being a Monday morning before coffee.

Anyway,  your proposal makes a lot more sense now. I'd say go with it. We can always add additional adapter identifier parameters later should the requirement arise. But the basic ones you've specified would work for most cases.

Is it worth adding the adapter enable/disable properties in this resource or would it be better creating an entirely different resource for configuring adapter properties? Although enable/disable is the most common scenario there are others that could be exposed down the track. @travisez13, @tysonjhayes?

If this resource did include the enable/disable parameter then the name should become xnetadapter I think.

Anyway, sorry for my confusion.

**Comment by gbrayut** [comment]
No worries. As for the enable/disable I can go either way. We don't currently name our disabled adapters (see screenshot above) so having to use two resources (one to set the name and one to disable) seems excessive. But I also see the benefit of a dedicated resource for setting adapter settings. Our current resource actually uses Ensure=Absent to indicate the specified network device should be disabled.

I am open to suggestions/recommendations either way. Our long term plan is managing MTU and other NIC settings, so either approach would be fine.

**Comment by TravisEz13** [comment]
@gbrayut, my problem is the proposed schema does not solve the problem as described in this issue, 

> rename a network adapter based on specific query criteria 

but instead, it lets you enable or disable the network adapter.   Enabling and disabling an adapter is valid functionality, but should be based on the alias, as the OS treats that as the key for the adapter.  The device name is difficult to predict ahead of time (binding order affects this which can be affected by many factors.)

**Comment by PlagueHO** [comment]
@TravisEz13  - I think this might work. Example:
```powershell
Configuration RenameAdapter {
    Import-DscResource xNetworking

    xNetAdapterName Adapter1 {
        InterfaceAlias = 'NewAdapterName'
        InterfaceDescription = 'Hyper-V Virtual Ethernet Adapter #1'
        MacAddress = '9C-D2-1E-61-B5-DA'
    }
}
```
The above in theory would rename the adapter with the MacAddress '9C-D2-1E-61-B5-DA' and description 'Hyper-V Virtual Ethernet Adapter #1'.

However, I'd actually argue that the parameter name 'InterfaceAlias' should actually be 'Name'.

It might help to actually list out the proposed MOF here - that would have made it clearer because it would show that the InterfaceAlias should not be a Key.

If it makes things clearer then I'd say ignore the State parameter for now.

**Comment by TravisEz13** [comment]
@PlagueHO For the behavior you describe I think the following scheme would work:
```
[ClassVersion("1.0.0"), FriendlyName("xNetAdapterName")]
class MSFT_xNetAdapterName : OMI_BaseResource
{
    [Key, Description("Desired name for the network adapter.")] string Name;
    [write, Description("Device description of the network adapter.")] string InterfaceDescription;
    [write,Description("Device MAC address")] string MacAddress;
};
```
You would have to add code in the resource to say at least one of the description or MAC must be supplied.

**Comment by PlagueHO** [comment]
@TravisEz13  - yep - that looks pretty good to me. I reckon we'd want to leverage the xNetworkAdapter "helper" module that you wrote a while back. That module could be extended to take the InterfaceDescrption and MacAddress parameters. It already takes the name (so that it can be set).

Seems like it might be a fairly straight forward process (famous last words).

**Comment by TravisEz13** [comment]
@PlagueHO , I think I agree.  The schema will still need to be reviewed by two other reviewers. @kwirkykat

**Comment by bcwilhite** [comment]
@TravisEz13 @kwirkykat Is there active work on this resource, if not, I'd like to help.  However, I have some concerns that I'd like to discuss.

Let me first start off by highlighting the need for InterfaceAlias renaming, for everyone's understanding:
_The need to have a specific NIC configured with the correct InterfaceAlias and therefore subsequent configuration activity, between OS installations, is due to physical network topology.  As an example one might have iSCSI NICs connected to an iSCSI network fabric, and production NICs connected to a separate network fabric, hence the need to ensure that the InterfaceAlias is named correctly, for the given physical NIC and/or virtual NIC, every time for a given physical host, regardless of the initial name the OS gives it when it's deployed._

My concerns with the discussed approach:

1. When thinking through this, we typically associate the **MAC Address** as being a unique property, however, when an LBFO Team is defined, you will have two MAC Addresses that will match, one for the **Primary** physical adapter and one for the **Team** virtual adapter.  So we may have to consider another unique property such as **Interface Alias** and/or **Interface Description**, which I'll highlight another concern in point two.
2. When installing the OS over and over on the same hardware, as typically done in test/dev environments, it is not guaranteed that you will have the same **Interface Alias** and **Interface Description** from installation to installation.  For example, if **Interface Alias** = **Ethernet**, **Interface Description** = **Intel(R) I210 Gigabit Network Connection** and **MAC Address** = **BC-5F-F4-C9-A0-95**, when the OS is redeployed, **"Ethernet"** may not be **"Intel(R) I210 Gigabit Network Connection"**, and the MAC may not equal **"BC-5F-F4-C9-A0-95"** as it did before.  So relying on these properties to be consistent between OS installations isn't viable, so we need to consider another option.

My initial thought is to require two properties, **DriverDescription** and **MACAddress**, to filter for a specific NIC, which should be consistent between OS deployments on the same hardware.

- **DriverDescription:** The driver name used for the given Physical and/or Virtual NIC
i.e. **"Intel(R) I210 Gigabit Network Connection"**

- **MacAddress:** The MAC Addressed assigned to the Physical and/or Virtual NIC

Having these two properties to filter for a specific NIC should suffice, since in theory the same Driver Description should not have two or more identical MAC addresses at all, especially in the same physical hosts.  My proposed schema would be as follows:

```
[ClassVersion("1.0.0"), FriendlyName("xNetAdapterName")]
class MSFT_xNetAdapterName : OMI_BaseResource
{
    [Key, Description("Desired name for the network adapter.")] string InterfaceAlias;
    [Required, Description("Device driver descripton of the network adapter.")] string DriverDescription;
    [Required, Description("Device MAC address of the network adapter.")] string MacAddress;
};
```

Notice the change in **InterfaceAlias** to be consistent with the rest of the xNetworking module resources.

I have other thoughts on this, but figured this is TL;DR already. ;-)

**Comment by TravisEz13** [comment]
@bcwilhite In the case of a virtual machine knowing the MacAddress ahead of time would be a very high obstacle, to the point of making the resource unusable for that scenario.

I'm open to making the driver description required.

**Comment by bcwilhite** [comment]
@TravisEz13 True, it would be difficult, my only concern is that you're able to nail down the exact adapter every time, so having the MacAddress at least optional would provide the facility for those that wanted to use it.  So in this case, as the user, they will have to understand their specific needs and what will work for them.

**Comment by kwirkykat** [comment]
@PlagueHO @TravisEz13 @bcwilhite @gbrayut 
I noticed there's already a xNetworkAdapterBinding resource. This can't be added to that resource with a DeviceName/Description property and/or MacAddress property?

Also, I'm not sure why this new resource would be called 'xNetworkAdapterName' rather than creating just  a 'xNetworkAdapter' resource?

Either way, I agree with Travis's proposed schema. Description and MacAddress should be optional to provide the flexibility to either provide both as @bcwilhite would need or to provide only one as would be needed in a virtual/testing environment.

Though I would suggest renaming Name --> InterfaceAlias (seems more consistent with the language here) and InterfaceDescription --> Description (seems to be an actual property of the adapter rather than just part of the interface as the alias is) as well as adding a readable Ensure property for Get-TargetResource to tell the user if the NetworkAdapter with the given name is Present or Absent.

**Comment by PlagueHO** [comment]
@kwirkykat , @TravisEz13 , @bcwilhite , @gbrayut 

It was a while ago since I proposed this resource, but from memory I suggested the name xNetworkAdapterName to differentiate it from the helper module xNetworkAdapter.psm1 @TravisEz13 had created for selecting an adapter. I'm not sure if having both the resource and the support module sharing the same name would end up causing us problems, but I think we should keep the names different just in case.

But renaming the support module wouldn't be that difficult because I don't think any of the resources actually use it yet. I think the original intent of xNetworkAdapter.psm1 was that any of the xNetworking resources that needed to specify an adapter could be retrofitted to use this module to select the adapter based on specified properties (mac address, driver name description etc). But so far none have been updated. I think it is possible that the only resource module that will use xNetworkAdapter.psm1 is this one (xNetworkAdapterName or whatever it will be called).

@kwirkykat , I originally suggested InterfaceAlias instead of Name because that would then match the InterfaceAlias name that was used in other resources that required an adapter to be specified. However, I don't think that is critical - I thought it just made it more obvious that the value that was being changed was actually this one.

Regarding the description parameter, I did a few queries on some of my servers and desktops and on every adapter on all of them the description parameter is empty. So I'm not sure about using the description parameter as an identifier of the adapter - it seems like it might not be useful.

Some other optional parameters that would also probably be useful as selectors would be InterfaceIndex and InterfaceName.

I think adding the Ensure parameter is a good idea as long as it is a Read parameter - otherwise people might think that they could add or remove adapters with this resource and try and set it.

A couple of edge case behaviors that need to be defined as well:
1. what should happen if an adapter matching the selectors can not be identified?
2. what should happen if more than one adapter matching the selectors is identified?

Sorry about the brain dump, but it has been a while since I was looking into this one so this is as much to refresh my own memory as anything.

**Comment by bcwilhite** [comment]
@kwirkykat I believe that DriverDescription is still needed, this property isn't as specific as InterfaceDescription, however, the combination of MacAddress and DriverDescription will give the user the flexibility of identifying the correct NIC in the event that the specific InterfaceDescription number is unknown.  DriverDescription is the Driver Name description, whereas InterfaceDescription is the Driver Name with an appended instance number given to the NIC at the time of OS Deployment, if there are multiple adapters with the same DriverDescription.
![image](https://cloud.githubusercontent.com/assets/12042836/21683465/7a11eb46-d327-11e6-8caf-3789ba7aa210.png)
Chances are that the user will know the DriverDescription, but not the InterfaceDescription "Instance" as depicted in the screenshot above.  My point is to include both, to provide the flexibility in that scenario.

**Comment by gbrayut** [comment]
Our preference is using InterfaceDescription to identify the NIC since it is stable across re-provisioning a system and is always unique. There can be some issues with identifying the value prior to OS Deployment, and occasional issues with re-numbering when moving NICs to a different PCI slot or USB port (requiring removing the old "hidden" devices from Device manager to go back to the original numbering), but we have found it to be the most stable identifier in our environment. For us it is also easier to manage than using MAC addresses, since we standardize our server setups (including NIC models and physical port connections). Also MAC addresses are not guaranteed to be unique in some advanced configurations as bcwilhite mentioned. 

Using the MAC address is our second preference, and I know some people prefer it because it is usually known before initial boot/install (often listed on the physical device itself). For VMs both VMWare and Hyper have options to identify or override the MAC address prior to installing the OS, but this may not be the case for all cloud providers. It also can change in some situations, like user overrides, VM migrations, or re-provisioning a VM on a new host.

Using DriverDescription is our least preferred option, as it will always require a second identifier if there is more than one instance in a system, which is fairly common in both VMs and physical servers. In all cases that second identifier (InterfaceIndex, InterfaceDescription, MAC) is much more unique than the DriverDescription.

So basically there are edge cases with all available options, and we likely need to be a bit flexible on the identifier used in this resource. The other xNetworking resources use the NIC Alias as the identifier since the OS requires it to be a unique value, which means as long as those resources include a dependency on the resource that sets the Alias they should avoid the above mentioned complications.

**Comment by TravisEz13** [comment]
@gbrayut I agree different users may have different preferences on how to find the adapter.  I'm not sure I agree that your statement of a guarantee about the Interface description is actually a guarantee by the operating system.

I think this leads to not requiring any of the fields other than the alias to be set. 

Going back and answering some of @PlagueHO 's questions I missed:
1.  If the resource does not find an adapter, it should error.  You are saying there should be an adapter meeting these conditions.
1. If there are multiple adapters, it should error, but I think this is a likely case.  I think the resource should have the option of using the first adapter.  

All of this would be described like this.

```
[ClassVersion("1.0.0"), FriendlyName("xNetAdapterName")]
class MSFT_xNetAdapterName : OMI_BaseResource
{
    [Key, Description("Desired name for the network adapter.")] string Name;
    [write, Description("Device description of the network adapter.")] string InterfaceDescription;
    [write, Description("Device driver descripton of the network adapter.")] string DriverDescription;
    [write,Description("Device MAC address")] string MacAddress;
    [write,Description("Action to take if the multiple adapters match the criteria.  Defaults to Fail."), ValueMap{'Fail','UseFirst'},Values{'Fail','UseFirst'}] string MultilpeMatchAction;
};
```

**Comment by PlagueHO** [comment]
@TravisEz13 - that looks pretty good to me.

We could also add the "Ensure" read parameter as well as @kwirkykat suggested, but I think the name isn't correct if it is just a read parameter. I'd suggest "Exists" or "IsFound":
```
[ClassVersion("1.0.0"), FriendlyName("xNetAdapterName")]
class MSFT_xNetAdapterName : OMI_BaseResource
{
    [Key, Description("Desired name for the network adapter.")] string Name;
    [write, Description("Device description of the network adapter.")] string InterfaceDescription;
    [write, Description("Device driver descripton of the network adapter.")] string DriverDescription;
    [write, Description("Device MAC address")] string MacAddress;
    [write, Description("Action to take if the multiple adapters match the criteria.  Defaults to Fail."), ValueMap{'Fail','UseFirst'}, Values{'Fail','UseFirst'}] string MultilpeMatchAction;
    [read, Description("Reports if the network adapter meeting the selection criteria is able to be found.")] bool Exists;
};
```

**Comment by TravisEz13** [comment]
@kwirkykat 
IRT, xNetworkAdapterBinding, I single adapter can have multiple bindings so combining the two would lead to a very complex resource.

IRT, this resource vs xNetworkAdapter, It would make at least the MacAddress complicated.  You would need two MacAddress properties.  One to find the adapter and one to set the adapters MacAddress.  To me, it makes more sense to separate the resource of identifying and setting the key for the resource and modifying the resource.  If we had [this user voice item in DSC](https://windowsserver.uservoice.com/forums/301869-powershell/suggestions/13399734-returning-information-from-dsc-resource-to-another ),  ideally this resource would not set anything but just ensure the resource is present and return the name.  Until then, I think we should act like setting the key is the act of returning it to the other resources.  Let's discuss offline.  Others on the thread feel free to comment.

**Comment by matthitchcock** [comment]
hi

I have a DSC Resource I am using internally for this task at the moment, in our scenario using the MAC address only as an identifier has been fine. The use case has been for Physical Hyper-V servers, we rename specific NIC's on specific networks so we do need to get the MAC addresses first. This information is given when the server is delivered anyway so has never been an issue.

Happy to contribute to this resource if needed, I want to get away from the internal module we are using.

**Comment by PlagueHO** [comment]
Hi @matthitchcock - any help is gratefully welcomed! I think we've all been talking about doing this resource for a while now, but nothing has actually happened. I put a forward a MOF spec for it a few comments back but I didn't get an "OK" on it from the community - @TravisEz13, @kwirkykat, @tysonjhayes ?

If we can get a "go" from them, then I'm definitely grateful for the help :grin:

**Comment by kwirkykat** [comment]
@PlagueHO After a long discussion with @TravisEz13 I understand why this resource is written this way since DSC doesn't have any way to retrieve the name of the NetAdapter. I'm fine with the latest schema proposal from @PlagueHO

**Comment by matthitchcock** [comment]
Just catching up on this thread, need to read and digest to understand the work involved. For now, I am getting by on my project using the resource I uploaded over the weekend.

I'll check out the Schema - would it require an OS to be already deployed onto the server to get the details required to set the NIC Alias name? I have been primarily considering bare metal scenarios so far.

**Comment by TravisEz13** [comment]
It would require that you have an OS deployed to run the configuration (which is always the case)  but not to set the NIC Name.

**Comment by PlagueHO** [comment]
This can be closed with the release of 4.1.0.0.
